Port to 1.21.5

Pixaurora opened the pull request:
So far, I've gotten everything to at least compile in 1.21.4 except for two modules: resource loader & registry.

@maloryware did more than half the work here, so I would like to thank him for all the time he put in!

Edit: Another big thanks to @supersaiyansubtlety and @alikindsys for their time contributing to this since they did stuff after I first made the draft PR! Also, it might be nice if you don't squash so it's easier to port, as I've already begun that.

Pixaurora committed:
Minecraft 1.21.4

Pixaurora committed:
Update access wideners

maloryware committed:
fixed a handful of mixin signatures, updated versions to target 1.21.4

maloryware committed:
updated `ItemStack::damage` signature to match new mappings

Pixaurora committed:
Fix some mapping names that changed

Pixaurora committed:
Update milk drinking mixin target

maloryware committed:
progress on ArmorFeatureRendererMixin fixing

Pixaurora committed:
Bump gradle and quilt loom versions

maloryware committed:
fixed multiple javadoc references (+ some mixins)

maloryware committed:
too tired to come up with a witty line for this commit. i fixed a few errors ig idfk anymore i can't bear mojang's codebase

maloryware committed:
sigh

maloryware committed:
refactored usages of DimensionTransition in QuiltDimensions

Pixaurora committed:
Hopefully fix ThreadedChunkManagerMixin

maloryware committed:
fixed a few mixins, annotated others with FIXME along w/ relevant info

Pixaurora committed:
Fix most compiler errors in entity_events

Pixaurora committed:
Update quilt client commands

maloryware committed:
fixed NetherBiomeSourceParameterMixin, annotated remaining mixins

maloryware committed:
Merge remote-tracking branch 'rina/1.21.4' into 1.21.4

alikindsys committed:
Document AbstractFurnaceBlockEntityMixin

alikindsys committed:
Add possible fixes for AbstractFurnaceBlockEntityMixin

alikindsys committed:
NBT: BurnTime -> lit_time_remaining

https://www.minecraft.net/en-us/article/minecraft-1-21-4-pre-release-2

Pixaurora committed:
Fix custom bows & screen render tests

Pixaurora committed:
Fix EndCrystalMixin in crash test mod

Pixaurora committed:
Fix everything except Registry & Resource Loader

Pixaurora committed:
Convert ResourceMetadataSectionReader usage to MetadataSectionType

alikindsys committed:
Fix AbstractFurnaceBlockEntityMixin

alikindsys committed:
Use `WrapOperation` instead of `ModifyVariable`

alikindsys committed:
Remove `Debug` annotation

Oops! That wouldn't have worked either way. Kept that around trying to debug it and forgot to remove it before pushing.

Pixaurora committed:
Merge pull request #1 from alikindsys/mixin

Fix `AbstractFurnaceBlockEntityMixin`

Pixaurora committed:
Make ScrollableMultiTextWidget inherit MultilineScrollableWidget

This just happens to unlock an astounding 45 compiler
errors and this seems to mean that I will have a LOT
more errors to fix than I initially believed. Not the
most thrilled now

Pixaurora committed:
Remove BUILTIN_ITEM_MODELS from ResourceLoaderKeys & fix Holder::value usage

Pixaurora committed:
Fully fix and test resource loader

Pixaurora committed:
Bump Quilt Mappings to 1.21.4 version 4

Pixaurora committed:
Fix more classes in the registry module

Pixaurora committed:
Fix everything except stuff that needs to be mapped

Pixaurora committed:
Update Minecraft version to 1.21.5

Pixaurora committed:
Fix more block classes and client command test

Pixaurora committed:
Port [& break] QuiltEntityTypeBuilder, and other various things

Pixaurora committed:
Remove 'Carver' support (API break) from QSL as it was similarly removed from the game

Pixaurora committed:
Remove ItemColor/ItemModel syncing since that is data-driven now

Pixaurora committed:
Quilt Mappings Version 6

Pixaurora committed:
Almost entirely port Quilt Testing

This almost entirely works, except for the fact that
Quilt's "empty" test structure is not loaded for
some reason. That will be fixed in Resource Loader

Pixaurora committed:
Load game test structures again, by updating the path searched

Pixaurora committed:
Update simple signature changes in Recipe module

maloryware commented:
> So far, I've gotten everything to at least compile in 1.21.4 except for two modules: resource loader & registry.
> 
> @maloryware did more than half the work here, so I would like to thank him for all the time he put in!

daw i just fixed signatures and left my input in as comments where i could, it's been really enjoyable working with you on this so far <3 keep up the golden work rina

supersaiyansubtlety committed:
remove VillagerType constructor accesswidener because VillagerType's only contructor is now public

supersaiyansubtlety committed:
remove TestFunction accesswidener because it's currenly unmapped and it no longer requires an accesswidener to be extensible

supersaiyansubtlety committed:
update type for WANDERING_TRADER_TRADES accesswidener

supersaiyansubtlety committed:
add remap=false to failure method targets in RegistryLibDynamicRegistryTest

supersaiyansubtlety committed:
update RegistryLibDynamicRegistryTest

supersaiyansubtlety committed:
update RegistryLibSyncOrderTest

supersaiyansubtlety committed:
update RegistryLibEventsTest, RegistryLibMonitorRegistrationTest, and RegistryLibMonitorTest

supersaiyansubtlety committed:
update RegistryLibSyncTest

supersaiyansubtlety committed:
check for GameTest annotation before accessing it

supersaiyansubtlety committed:
add missing lateInit call to ClientConfigurationNetworkHandlerMixin::initAddon

supersaiyansubtlety committed:
fix RuntimeException 'Registry snapshot already exists!' crash when tryint to connect after certain dsiconnects

supersaiyansubtlety committed:
add quilt$ prefix to ChannelInfoHolder::getPendingChannelsNames (implemented by mixin)

supersaiyansubtlety committed:
fix misleading mixin member name

supersaiyansubtlety committed:
Revert "add missing lateInit call to ClientConfigurationNetworkHandlerMixin::initAddon"

This reverts commit a130f62faff0b5d46297f2546883483ff0d05bb7.

supersaiyansubtlety committed:
rework onBackupExperimentalWarning so it doesn't copy/paste a lambda

Pixaurora committed:
Merge pull request #2 from supersaiyansubtlety/1.21.4

registry test progress

supersaiyansubtlety committed:
update WorldSaveStorageBugfixMixin

Pixaurora committed:
Port Block Content Registries

supersaiyansubtlety committed:
update mappings

supersaiyansubtlety committed:
update bow attack modification

supersaiyansubtlety committed:
entity_rendering testmod launches but armor is untextured
breaking changes to ArmorRenderingRegistry

supersaiyansubtlety committed:
improve mixin hygiene

supersaiyansubtlety committed:
add gui:screen to item_extension testmod deps
update item_extension test items

supersaiyansubtlety committed:
add missing BOW_MODIFY_SHOT_PROJECTILE invocation to RangedWeaponItemMixin

supersaiyansubtlety committed:
move RangedWeaponItemMixin::modifyArrow implementation to BowItemMixin so modded non-bow ranged weapons aren't affected

supersaiyansubtlety committed:
enter the skeleton cheat code, hand pitchforks to the illusioners

Pixaurora committed:
Merge pull request #3 from supersaiyansubtlety/bow-attack-modification

Bow attack modification

supersaiyansubtlety committed:
add missing original call to ArmorFeatureRendererMixin::quilt$modifyRenderLayer
make ArmorRenderingRegistryImpl::getArmorRenderLayer return @NotNull
minor cleanup

supersaiyansubtlety committed:
fix resource locations

supersaiyansubtlety committed:
separate quilt_leggings equipment asset and point to it in client testmod

supersaiyansubtlety committed:
replace quilt_helmet model with witch head

supersaiyansubtlety committed:
initial pass at recipe update

Pixaurora committed:
fix recipe API and test mod signatures

Pixaurora committed:
Merge pull request #4 from supersaiyansubtlety/armor-rendering

Armor rendering

supersaiyansubtlety committed:
add RecipeData and implement ShapedRecipeData

supersaiyansubtlety committed:
add ShapedRecipeData.Builder

supersaiyansubtlety committed:
refine and javadoc ShapedRecipeData

supersaiyansubtlety committed:
fix javadoc typo

supersaiyansubtlety committed:
got RecipeTestMod's static recipe with tag input working with the new ShapedRecipeData

supersaiyansubtlety committed:
add id to recipe creation error message

supersaiyansubtlety committed:
removed RecipeBuilders as they fail to build recipes with tags
replaced usage of RecipeManagerImpl.currentRegistryManager with RecipeManager.registries
updated recipe add, remove, and replace events
fix RecipeManagerMixin::applyModifications applying too late
fix RemoveRecipeHandlerImpl::removeIf(RecipeType<T>, Predicate<RecipeHolder<T>>) not removing recipes from byType

supersaiyansubtlety committed:
implement ShapelessRecipeData
replace Either with IngredientData in RecipeData implementations

supersaiyansubtlety committed:
javadoc ShapelessRecipeData

supersaiyansubtlety committed:
add VanillaRecipeData, remove VanillaRecipeBuildersImpl

supersaiyansubtlety committed:
javadoc VanillarecipeData

supersaiyansubtlety committed:
javadoc IngredientData

supersaiyansubtlety committed:
copy RecipeData result stacks

Pixaurora committed:
Merge pull request #5 from supersaiyansubtlety/recipe

Recipe

supersaiyansubtlety committed:
reduce visibility of several VanillaRecipeData constants

supersaiyansubtlety committed:
initial pass at trade offer api rework, theoretically complete except for the test mod
flatten TradeOfferHelper.WanderingTraderOffersBuilder functionality into static methods, assisted by WanderingTraderOffersManager
split villager trade adding into separate methods for default and experimental trades
replace usage of VillagerOffersAdder and Comsumer<List> params with vararg and Collection params

supersaiyansubtlety committed:
remove unnecessary accessors, add @Mutable to WANDERING_TRADER_TRADES setter

supersaiyansubtlety committed:
update remaining entity_extension mixins
update entity_extensions test mods+mixins

Pixaurora committed:
Merge pull request #6 from supersaiyansubtlety/recipe-constants

reduce visibility of several VanillaRecipeData constants

Pixaurora committed:
Merge pull request #7 from supersaiyansubtlety/trades

trade offers

Pixaurora committed:
Fix trivial signature changes in item content registry

Pixaurora committed:
port item content registry to 1.21.5

supersaiyansubtlety committed:
update item_settings testmod data

supersaiyansubtlety committed:
remove EquipmentSlotProvider and QuiltItemSettings::equipmentSlot as they're redundant with vanilla's EquippableComponent
add new Overrides of vanilla methods to QuiltItemSettings

supersaiyansubtlety committed:
update remaining item_settings things

supersaiyansubtlety committed:
minor cleanup

supersaiyansubtlety committed:
clean up item_settings mixins

supersaiyansubtlety committed:
fix BrewingStandBlockEntityMixin breaking all brewing

supersaiyansubtlety committed:
fix recipe damage+self remainders copying input stack count

supersaiyansubtlety committed:
death to horizontal scrolling

Pixaurora committed:
Merge pull request #8 from supersaiyansubtlety/item_settings

Item settings

supersaiyansubtlety committed:
fix AFTER_ENTITY_WORLD_CHANGE mixin

supersaiyansubtlety committed:
init modded blocks' states' caches (caused crash when rendering a modded block)

supersaiyansubtlety committed:
update QuiltBlockEntityTypeBuilder

supersaiyansubtlety committed:
update BlockEntityTypeTest

supersaiyansubtlety committed:
update PredicateCodecMap javadoc reference

supersaiyansubtlety committed:
update FUEL_FILTERS

supersaiyansubtlety committed:
fix test composter data

supersaiyansubtlety committed:
update populate/bindTag

supersaiyansubtlety committed:
fix fixmes in effect LivingEntityMixin

Pixaurora committed:
Merge pull request #9 from supersaiyansubtlety/entity_events+block_entity-fixes

`entity_events` and `block_entity` fixes

supersaiyansubtlety committed:
update more tags module test data
replace SimpleRegistryMixin with TagGroupLoaderMixin (equivalent, but accounts for future non-SimpleRegistry MutableRegistry implementations)
add SimpleRegistryPendingTagsMixin to capture tags for more registries

supersaiyansubtlety committed:
update StatusEffectTest

supersaiyansubtlety committed:
change whitespace in ArmorFeatureRendererMixin

supersaiyansubtlety committed:
fix TooltipTestMod

supersaiyansubtlety committed:
fix registry_entry_attachment test mods

Pixaurora committed:
Merge pull request #10 from supersaiyansubtlety/fixmes

Fix FIXMEs

supersaiyansubtlety committed:
fix block_extensions testmod

supersaiyansubtlety committed:
fix example_biome.json, add FIXME to QuiltBiomeTest for biome colors

supersaiyansubtlety committed:
sync modified biomes from known packs to client

supersaiyansubtlety committed:
replace EntityPart:renderHitbox with getHitbox
replace EntityRenderDispatcherMixin with EntityRendererMixin
fix ThreadedChunkManagerMixin

Pixaurora committed:
Merge pull request #11 from supersaiyansubtlety/fix-tests

Fix tests

Pixaurora committed:
use optional packet codec in-case the random item was air

OroArmor commented on library/core/testing/src/main/java/org/quiltmc/qsl/testing/api/game/annotation/GameTest.java:
```diff
@@ -0,0 +1,67 @@
+package org.quiltmc.qsl.testing.api.game.annotation;
```
No license header

OroArmor commented on library/core/testing/src/main/java/org/quiltmc/qsl/testing/api/game/annotation/GameTest.java:
```diff
@@ -0,0 +1,67 @@
+@Deprecated
+@Target({ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface GameTest {
```
Was the minecraft annotation removed so this is a placeholder for now to make migration easier?

Pixaurora replied:
Yes, although maybe it should just be an option we provide, as code-driven tests make more sense to me than data-driven ones

OroArmor commented on library/core/testing/src/main/java/org/quiltmc/qsl/testing/mixin/TestServerMixin.java:
```diff
@@ -38,7 +38,8 @@ public void isDedicated(CallbackInfoReturnable<Boolean> cir) {
 			constant = @Constant(stringValue = "All {} required tests passed :)")
 	)
-	private static String replaceSuccessMessage(String original) {
+	// was static, signature didn't match target method. target method signature probably changed
```
Comment can be removed

OroArmor commented on library/data/registry_entry_attachment/src/testmod/java/org/quiltmc/qsl/registry/attachment/test/AttachmentTestUtil.java:
```diff
@@ -0,0 +1,45 @@
+
+public final class AttachmentTestUtil {
+	private AttachmentTestUtil() {
+		throw new AssertionError(AttachmentTestUtil.class.getSimpleName() + " contains only static members");
```
I think we usually used illegal state exception or something else. I would look at other static classes.

supersaiyansubtlety replied:
```suggestion
		throw new UnsupportedOperationException(AttachmentTestUtil.class.getSimpleName() + " contains only static members");
```

OroArmor commented on library/entity/entity_extensions/src/main/java/org/quiltmc/qsl/entity/extensions/api/MinecartComparatorLogic.java:
```diff
@@ -23,6 +23,8 @@
 
 import org.quiltmc.qsl.base.api.util.InjectedInterface;
 
+// should the DetectorRailBlock#getComparatorOutput reference error here??
```
Could you elaborate more?

Pixaurora replied:
it's because the method has `protected` access, but I checked and it also is on the 1.21.1 branch so I think this isn't a problem

OroArmor commented on library/entity/status_effect/src/main/java/org/quiltmc/qsl/entity/effect/mixin/ClearAllEffectsComponentMixin.java:
```diff
@@ -16,22 +16,24 @@
+public abstract class ClearAllEffectsComponentMixin {
+	@Inject(method = "apply", at = @At(value = "HEAD"))
+	private void quilt$addRemovalReason(World world, ItemStack stack, LivingEntity entity, CallbackInfoReturnable<Boolean> cir) {
+		entity.clearStatusEffects(StatusEffectRemovalReason.DRANK_MILK);
```
Since we no longer know this is milk, should we change the name of the removal reason?

Pixaurora replied:
I asked that before and was told no, ultimately it's still 'mechanically' milk so is it that different?

OroArmor replied:
Maybe we could leave the drank milk option if the stack fits a milk bucket tag (im thinking about milk bottles and such), otherwise it would default to a more default option. We could also maybe add our own component that specifically says what the clearing type is, since we cant really attach it directly to the component itself

OroArmor commented on library/item/item_content_registry/src/main/java/org/quiltmc/qsl/item/content/registry/mixin/MinecraftServerMixin.java:
```diff
@@ -0,0 +1,21 @@
+
+@Mixin(MinecraftServer.class)
+public class MinecraftServerMixin {
+	@Inject(method = "<init>", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/entity/FuelTimes;create(Lnet/minecraft/registry/HolderLookup$Provider;Lnet/minecraft/feature_flags/FeatureFlagBitSet;)Lnet/minecraft/block/entity/FuelTimes;"))
+	private void startFuelTimeCollection(CallbackInfo ci) {
+		ItemContentRegistriesInitializer.startInitialFuelCollection();
+	}
+
+	@Inject(method = "<init>", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/entity/FuelTimes;create(Lnet/minecraft/registry/HolderLookup$Provider;Lnet/minecraft/feature_flags/FeatureFlagBitSet;)Lnet/minecraft/block/entity/FuelTimes;", shift = At.Shift.AFTER))
+	private void endFuelTimeCollection(CallbackInfo ci) {
+		ItemContentRegistriesInitializer.endInitialFuelCollection();
+	}
```
Id do a wrap operation and call the original in between these two calls. Makes it clearer what the mixins are doing

OroArmor commented on library/item/item_extensions/src/main/java/org/quiltmc/qsl/item/extensions/mixin/bow/BowItemMixin.java:
```diff
@@ -21,17 +21,48 @@
+			// speed is calculated from pullProgress * 3 in BowItem::onStoppedUsing
+			final float pullProgress = speed / 3f;
+
+			final PersistentProjectileEntity projectile = BowShotProjectileEvents.BOW_REPLACE_SHOT_PROJECTILE.invoker()
```
Call `BowAttackModiciationImpl#modifyShotProjectile` here?

OroArmor commented on library/item/item_setting/src/main/java/org/quiltmc/qsl/item/setting/api/QuiltItemSettings.java:
```diff
@@ -190,6 +183,7 @@ public <T> QuiltItemSettings customSetting(CustomItemSetting<T> setting, T value
 	}
 
 	// Overrides of vanilla methods
+	// TODO override remaining methods once they're mapped
```
I think it would be better to make this an injected interface, then we dont have to worry about forgetting to make overrides for the different settings.

OroArmor commented on library/item/item_setting/src/main/java/org/quiltmc/qsl/item/setting/api/RecipeRemainderLogicHandler.java:
```diff
@@ -50,12 +50,15 @@ public interface RecipeRemainderLogicHandler {
+			CustomItemSettingImpl.RECIPE_REMAINDER_PROVIDER.get(original.getItem());
 
-		RecipeRemainderProvider provider = (_original, _recipe) -> _original.getItem().hasRecipeRemainder() ? _original.getItem().getRecipeRemainder().getDefaultStack() : ItemStack.EMPTY;
+		RecipeRemainderProvider provider = (_original, _recipe) -> _original.getItem().getRecipeRemainder();
```
Is there a reason this was modified? I'm fairly certain I added this to fix a bug

supersaiyansubtlety replied:
`hasRecipeRemainder` doesn't exist anymore, but `getRecipeRemainder` is `final` and returns `ItemStack.EMTPY` if the remainder is `null`

OroArmor commented on library/item/item_setting/src/main/java/org/quiltmc/qsl/item/setting/mixin/recipe_remainder/LoomOutputSlotMixin.java:
```diff
@@ -21,20 +21,27 @@
 	}
 
-	@Redirect(method = "onTakeItem(Lnet/minecraft/entity/player/PlayerEntity;Lnet/minecraft/item/ItemStack;)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/screen/slot/Slot;takeStack(I)Lnet/minecraft/item/ItemStack;"))
+	// MCDev erroneously says this method and target are incorrect; the anonymous class is probably confusing it
```
yeah. mcdev doesnt like this

OroArmor reviewed (changes requested):
Initial review looks pretty good. There are a lot of missing licenses and checkstyle violations, I would make sure those are all fixed. Also I see the addition of a lot of `final` keywords. I'm not sure if that super needed, but I think its ok.

supersaiyansubtlety commented on library/item/item_extensions/src/main/java/org/quiltmc/qsl/item/extensions/mixin/bow/BowItemMixin.java:
```diff
@@ -21,17 +21,48 @@
 import net.minecraft.world.World;
 
 import org.quiltmc.qsl.item.extensions.api.bow.BowExtensions;
+import org.quiltmc.qsl.item.extensions.api.bow.BowShotProjectileEvents;
+
+import com.llamalad7.mixinextras.sugar.Local;
```
```suggestion
import org.quiltmc.qsl.item.extensions.impl.BowAttackModificationImpl;
```

supersaiyansubtlety commented on library/item/item_extensions/src/main/java/org/quiltmc/qsl/item/extensions/mixin/bow/BowItemMixin.java:
```diff
@@ -21,17 +21,48 @@
+			// speed is calculated from pullProgress * 3 in BowItem::onStoppedUsing
+			final float pullProgress = speed / 3f;
+
+			final PersistentProjectileEntity projectile = BowShotProjectileEvents.BOW_REPLACE_SHOT_PROJECTILE.invoker()
+				.replaceProjectileShot(
+					bowStack, arrowStack, user, pullProgress,
+					persistentProjectile
+				);
+
+			BowShotProjectileEvents.BOW_MODIFY_SHOT_PROJECTILE.invoker().modifyProjectileShot(
+				bowStack, arrowStack, user, pullProgress, persistentProjectile
+			);
+
+			return projectile;
```
```suggestion
			return BowAttackModificationImpl.modifyShotProjectile(
				persistentProjectile,
				arrowStack, pullProgress, bowStack, user
			);
```

supersaiyansubtlety commented:
Something weird I noticed when testing:
Running `gradlew :core:networking:check` runs a task that uses the `networking/` directory as a run directory, instead of `networking/run/`.

It creates `.cache/`, `config/`, `logs/`, and `mods/` directories under `networking/`.
Running `gradlew :core:networking:runGameTestServer --rerun-tasks` does *not* do this, so I'm not sure which task is doing it.

supersaiyansubtlety committed:
replace QuiltItemSettings with injected interface: QuiltItemSettingsExtensions

Pixaurora committed:
Merge pull request #12 from supersaiyansubtlety/inject-item-settings

Inject item settings

Pixaurora committed:
Remove checkItemSetting as it is now an injected interface

Pixaurora committed:
Add missing licenses

Pixaurora committed:
Apply supersaiyansubtlety's suggestions from code review

Co-authored-by: Will <supersaiyansubtlety@gmail.com>

Pixaurora committed:
Apply other recommendations from Oro

supersaiyansubtlety committed:
fix checkstyle and other cleanup

Pixaurora committed:
Merge pull request #13 from supersaiyansubtlety/fix-checkstyle

fix checkstyle

OroArmor reviewed (changes requested):
Pretty good. I think I wasn't clear enough on all of the final stuff for variables. I more meant that I don't think it is needed unless it is required to compile the code. It just adds extra words to read when looking at the code, which slows down reading. Obviously for fields it should be used as much as possible.

supersaiyansubtlety commented:
> Pretty good. I think I wasn't clear enough on all of the final stuff for variables. I more meant that I don't think it is needed unless it is required to compile the code. It just adds extra words to read when looking at the code, which slows down reading. Obviously for fields it should be used as much as possible.

My bad, I should have asked for clarification before adding more finals. I was hoping they’d increase readability since you don’t have to look for reassignment, you always know the value. 

Should I make another PR to remove the finals I added in the checkstyle PR?

OroArmor commented:
> My bad, I should have asked for clarification before adding more finals. I was hoping they’d increase readability since you don’t have to look for reassignment, you always know the value.

I know in intellij it will have an underline for non-final variable. At least i remember something like this. I know it does it with rust, would be surprised if it didnt for java.

Pixaurora commented:
it only underlines it if you edit the variable I think, with "reassigned variable"

Pixaurora committed:
update actions

supersaiyansubtlety committed:
revert finals from checkstyle PR

supersaiyansubtlety committed:
remove finals added since 1.21

Pixaurora committed:
Merge pull request #14 from supersaiyansubtlety/revert-finals

revert finals from checkstyle PR
